src/Streamer.cxx

```diff
@@ -1,321 +1,143 @@
-/// the given topic. If a start time is specified retrieve the correct initial
-/// log. Assign the TopicPartition vector to the Consumer
-/// \param TopicName the topic that the Streamer will consume
-/// \param Options a StreamerOptions object
-/// that contains configuration parameters for the Streamer and the KafkaConfig
-void FileWriter::Streamer::connect(const std::string &TopicName) {
-  std::lock_guard<std::mutex> lock(ConnectionReady);
-
-  std::lock_guard<std::mutex> lk(ConnectionLock);
-  Initialising = true;
-  ConnectionInit.notify_all();
+// pass the topic by value: this allow the constructor to go out of scope
+// without resulting in an error
+FileWriter::Streamer::SEC FileWriter::Streamer::connect(std::string TopicName) {
 
   LOG(Sev::Debug, "Connecting to {}", TopicName);
-  auto Config = createConfiguration(Options);
-  if (!Config) {
-    return;
-  }
-  RunStatus = createConsumer(std::move(Config));
-  if (RunStatus != SEC::no_error) {
-    return;
-  }
-  auto Metadata = createMetadata();
-  if (!Metadata) {
-    return;
-  }
-  LOG(Sev::Debug, "createMetadata");
-  RunStatus = createTopicPartition(TopicName, std::move(Metadata));
-  if (RunStatus != SEC::no_error) {
-    return;
-  }
-  LOG(Sev::Debug, "createTopicPartition");
-  RunStatus = assignTopicPartition();
-  if (RunStatus != SEC::no_error) {
-    return;
+  try {
+    Consumer.reset(new KafkaW::Consumer(Options.Settings));
+    if (Options.StartTimestamp.count()) {
+      Consumer->addTopic(TopicName,
+                         Options.StartTimestamp - Options.BeforeStartTime);
+    } else {
+      Consumer->addTopic(TopicName);
+    }
+    // Error if the topic cannot be found in the metadata
+    if (!Consumer->topicPresent(TopicName)) {
+      return SEC::topic_partition_error;
+    }
+  } catch (std::exception &Error) {
+    LOG(Sev::Error, "{}", Error.what());
+    return SEC::configuration_error;
   }
-  LOG(Sev::Debug, "Connected to topic {}", TopicName);
-  RunStatus = SEC::writing;
+
+  return SEC::writing;
 }
 
 FileWriter::Streamer::SEC FileWriter::Streamer::closeStream() {
-  std::lock_guard<std::mutex> lock(ConnectionReady);
-  if (ConnectThread.joinable()) {
-    ConnectThread.join();
-  }
-  if (Consumer) {
-    Consumer->close();
-  }
-  TopicPartitionVector.clear();
-  //  if (RunStatus == SEC::writing) {
+  Sources.clear();
   RunStatus = SEC::has_finished;
-  //  }
   return RunStatus;
 }
 
 template <>
 FileWriter::ProcessMessageResult
 FileWriter::Streamer::write(FileWriter::DemuxTopic &MessageProcessor) {
 
-  if (RunStatus == SEC::not_initialized) {
-    return ProcessMessageResult::OK();
+  // wait for connect() to finish
+  try {
+    if (IsConnected.valid()) {
+      if (IsConnected.wait_for(std::chrono::milliseconds(1000)) !=
+          std::future_status::ready) {
+        LOG(Sev::Critical, "... still not ready");
+        return ProcessMessageResult::OK();
+      }
+      std::call_once(ConnectionStatus,
```

Comment:
Why is `std::call_once` necessary here?

Reply:
Not anymore, before the fture::get was outside of the if condition. Fixed.